src/libutil/serialise.cc

```diff
@@ -158,6 +160,24 @@ bool FdSource::good()
 }
 
 
+bool FdSource::hasData()
+{
+    if (BufferedSource::hasData()) return true;
+
+    while (true) {
+        struct pollfd fds[1];
+        fds[0].fd = fd;
+        fds[0].events = POLLIN;
+        auto n = poll(fds, 1, 0);
+        if (n < 0) {
+            if (errno == EINTR) continue;
+            throw SysError("polling file descriptor");
+        }
+        return n == 1 && (fds[0].events & POLLIN);
+    }
```

Comment:
Can you use `select` instead of `poll` since windows has it? 
https://learn.microsoft.com/en-us/windows/win32/api/winsock2/nf-winsock2-select

Don't think it should matter match for a single FD.

Reply:
Done.

Reply:
Hm, ideally a better solution should be found than using `select`. Select *is* problematic even for a single file descriptor, because select can only work with low-numbered file descriptors. Eg, If you put just a single fd into select, but that fd happens to have large numeric value larger than 1024, then you'd get an error. See the WARNING section from the manpage: https://man7.org/linux/man-pages/man2/select.2.html

We hit this in cargo a while ago, when Cargo was spawned by a different process which didn't set CLOEXEC flag on its file descriptors, so fds cargo ended up seeing were beyond `FD_SETSIZE`: https://github.com/rust-lang/cargo/pull/4477